I1030 WTI clar announcement filtering

clevengr opened the pull request:
### Description of what the PR does

Fixes the WTI-UI clarification filter buttons so that they properly display both clarifications and announcements according to the selected filter button. Providing this support required updating the WTI clarification model, and also required providing support for the ability to send announcements to groups and teams simultaneously.  The PR includes fixes for all these, along with a number of related fixes and "continuous improvement" items, as follows:
- creates a new class `GenerateAnnouncementPane` supporting creation of announcements.  The new class, based on (and credit goes to) earlier work by Kutay Karakas, provides support for:
  - Separate simultaneous selection of Groups and Teams as announcement destinations.
  - Listener support for automatically updating group and team scrollpanes if the Admin makes group or team (account) edits.
- Updates the WTI-UI clarifications page for improved readability, including updating filter buttons.
- Fixes minor issues such as replacing `ItemStateListeners` with `ActionListeners` where appropriate.
- Adds support in WTI-API class `ContestController` to allow clarifications to be returned to the logged-in user any time, and to allow the problem list to be returned any time after the contest has been started (previously, the contest had to be _running_ to get clarifications or problems).
- Adds a new "Announce" pane to both the Judge and the Admin views, replacing (removing) the old `SubmitClarificationPane` which is no longer needed (because the sole reason for "submit clarification" was to create a "question" which could then be "answered" with an announcement).
- Reverts class `SubmitClarificationPane` to the version prior to when "announcements" were added to it (since announcements are now supported by a separate class).
- Adds the `SubmitClarificationPane` to the Developer pane (so that it is still available if needed).

### Issue which the PR addresses

Fixes #1030; also fixes #1039 and fixes #1024.

### Environment in which the PR was developed:

Windows 10 & 11; Eclipse 2020-12 (4.18.0); Java 1.8.0_271.

### Precise steps for _testing_ the PR (i.e., how to demonstrate that it works correctly)

#### Setup
- Download the PR distribution.
- Start a PC2 server using a contest with groups, teams, problems, and judges defined (e.g. `--load mini`).
- Start a PC2 Admin.
  - ensure there is a "Scoreboard2" account (create one if not)
  - start the contest running.
- Start a manual judging client (e.g. using `./bin/pc2judge --login j1`).
- Start the WTI Server (go into the "projects" folder and unzip "WebTeamInterface-1.x-.zip", then into the unzipped folder and execute "./bin/pc2wti").
- Start a browser; if you've been running PC2 in this browser then CLEAR THE BROWSER'S DATA CACHE.
- Point the browser at the PC2 WTI server (typically, "http://localhost:8080").

#### Test Normal Clars
- login as a team (if using the "mini" contest, use "team101").
- Select the "Clarifications" menu item.
- On the Clarifications page, click "New Clarification".
- Select a Problem (if you're using "mini", there's only one: "Sumit".  If you use the Admin to add problems, you should see more).
- Enter some text for the clarification, then click "Submit".
- On the Admin or Judge, select the "Clarifications" tab, select the new clarification, click the "Answer" button, and enter some "Answer text". Do NOT check "Send to all teams".  Hit "OK".
- On the team browser, Submit a second clarification.
- On the Adminn or Judge, answer the second clarification, but DO check the "Send to ALL teams" box before clicking "OK".
- On the team browser, Submit a third clarification -- but DO NOT answer it on the Admin or Judge.
- On the team browser, click the "Runs" menu and then click back to the "Clarifications" menu (this is necessary because the PR is based on a version of Develop that does not yet support
	dynamic updating of the Clarifications page).
- Verify that:
  - Clicking the "All Teams" radio button displays the clar whose reply was sent to All Teams and the clar which hasn't been answered yet (but no others).
  - Clicking the "Selected Teams" radio button displays just the unanswered clar (but NO other clars -- because no answers have yet been sent to SOME but not ALL teams).
  - Clicking the "My Team" radio button displays the clar which was answered just to team 1 (but NOT to all teams), and the unanswered clar -- but no others.
  - Clicking the "All the Above" button displays all clars, including the unanswered one.

#### Test Announcements
##### Announcements to All Teams
- On the judge or Admin, select the "Announce" tab then do the following:
  - Select a problem.
  - Select "All Teams" as the destination (it should be selected by default).
  - Enter some text in the "Announcement Text" box.
  - Click the "Send Announcement" button, then click "Yes" to confirm sending the Announcement.
- On the team browser, select the "Runs" page (menu item), then select the "Clarifications" page again.
- Verify that the Announcement shows up when the "All Teams" or the "All the Above" buttons are selected, but NOT when the "Selected Teams" or "My Team" buttons are selected.

##### Announcements to a Specific Team
- On the judge or Admin, select the "Announce" tab then do the following:
  - Select a problem.
  - Select "Specific Groups and/or Teams" as the destination.
  - In the "Teams" list, check the box for "team 101" and for "team 102" (or some set of two teams that exist in your contest).
  - Enter some text in the "Announcement Text" box.
  - Click the "Send Announcement" button, then click "Yes" to confirm sending the Announcement.
- On the team browser, select the "Runs" page, then go back to the "Clarifications" page.  
- Verify that the second Announcement shows up when the "Selected Teams" or the "All the Above" buttons are selected, but NOT when the "All Teams" or "My Team" button is selected.

##### Announcements to Groups (only)
- On the judge or Admin, select the "Announce" tab then do the following:
  - Select a problem.
  - Select "Specific Groups and/or Teams" as the destination.
  - In the "Teams" list, UNSELECT any teams which are selected.  (You can also do this by changing the "Destinations" dropdown to "All Teams" and then back to "Specific Teams/Groups"; this action clears all group and team selections).
  - In the "Groups" list, check the box for a group of which your logged-in team is a member.
  - Enter some text in the "Announcement Text" box.
  - Click the "Send Announcement" button, then click "Yes" to confirm sending the Announcement.
- On the team browser, select the "Runs" page, then go back to the "Clarifications" page.  
- Verify that the third Announcement shows up when the "Selected Teams" or the "All the Above" buttons are selected, but NOT when the "All Teams" or "My Team" button is selected.

##### Announcements to both Groups AND Specific Teams
- On the judge or Admin, select the "Announce" tab then do the following:
  - Select a problem.
  - Select "Specific Groups and/or Teams" as the destination.
  - In the "Groups" list, check the box for the group of which your logged-in team is a member.
  - In the "Teams" list, SELECT your logged-in team, and also select another team which is in the same group, and one team which is NOT in the selected group.
  - Enter some text in the "Announcement Text" box.
  - Click the "Send Announcement" button, then click "Yes" to confirm sending the Announcement.
- On the team browser, select the "Runs" page, then go back to the "Clarifications" page.  
- Verify that the fourth Announcement shows up when the "Selected Teams" or the "All the Above" buttons are selected, but NOT when the "All Teams" or "My Team" button is selected.

##### Test Another Team
- Open a new browser window.
- Clear the browser's cache.
- Enter the PC2 connect URL into the browser's address bar.
- Login as a different team (specifically, the second team selected in the above tests).
- Select the "Clarifications" page (menu item).
- Click the "All Teams" filter button; verify that the team sees all (and only) the clars and announcements sent to "All Teams".
- Click the "Selected Teams" filter button; verify that the team sees only clars and announcements sent specifically to this team or to a group of which this team is a member.
- Click the "My Team" filter button; verify that the team sees "No Clarifications Yet!"  (because this team has not submitted any clarifications).
- Click the "All the Above" filter button; verify that the clarification list shows the union of what the other three buttons show.

clevengr committed:
i1030: remove redundant "Recipient"  column from clars page.

clevengr committed:
i1030: set announcement clar state for empty OR "announce" question:

previously, announcement clars were only set to state "ANNOUNCED" if the
question was empty.  Empty questions are confusing on the team grid, so
announcements were updated to contain a "question text" value of
"Announcement".  This commit changes the setAnswer() methods to set the
state to "ANNOUNCED" if the Announcement Question text is EITHER empty
OR "announcement" (ignoring case).

clevengr committed:
i1030: remove test for "isJudge" in deciding clars to send to team.

clevengr committed:
i1030: set announcement clarification question text to "Announcement"

Commit:
i1030: add doc, refactor some names, update for announcements.

Commit:
i1030: add/update doc; refactor some names (no logic changes).

Commit:
i1030: update the way an intended clar recipient is specified.

Commit:
i1030: update doc to clarify what clars are returned.  No logic changes.

Commit:
i1030: replace combobox ItemListener with ActionListener, because...

...an "ItemListener" is invoked twice when a combobox changes (once for
the unselected item; again for the newly-selected item) whereas we're
only interested in the fact that an action selected a new item.

Also removed a redundant adding of groupsAndTeamsList to the groupsPanel
(it was being added twice but should only be added once).
Also cleaned up some documentation and layout sizes, and refactored some
variable and method names.

Finally, added a "TODO" comment on one place where the code doesn't make
any sense and it potentially VERY inefficient.

Commit:
i1030: change WTI-UI button label from "Just My Team" to "My Team".

Commit:
i1030: initial pass at separating Groups/Teams scrollpanes.

clevengr committed:
i1030: replace SubmClarPane with GenerateAnnouncementPane.

clevengr committed:
i1030: initial version of new GenerateAnnouncementPane.

clevengr committed:
i1030: minor additions to SubmiClarPane.  However:

this class should probably be reverted all the way back to its original
version, before any modifications were made to it for supporting
"announcements".  A completely new class, "GenerateAnnouncement", has
been implemented and added to the JudgeView pane to support
announcements; this class has been removed from the JudgeView -- it
should be added (in its original form) to the "Developer's Pane".

clevengr committed:
i1030: add SubmitClarificationPane to DeveloperPane.  Note:

the current implementation of SubmitClarificationPane throws an
exception when run from the DeveloperPane.  The SubmitClarificationPane
class needs to be reverted all the way back to what it was before
modifications were may to support "announcements" (that function is now
supported in new class "GenerateAnnouncement").

clevengr committed:
i1030: add GenerateAnnouncements pane to Admin View.

clevengr committed:
i1030: minor renaming and unnecessary code removal.

Commit:
i1030: rename tabs on JudgeView.

Commit:
i1030: remove unnecessary (commented-out) code.

Commit:
i1030: implement getGroupsAndTeamsSelectedValues()

Commit:
i1030: change tab labels from "Announcements" to "Announce".

Commit:
i1030: update logic to handle simultaneous groups and teams.

Commit:
i1030: reverted SubmitClarificationPane to earlier version, because...

changes made after the reverted version were all related to supporting
"generate announcement" capability, which has now been implemented in
separate class "GenerateAnnouncement".

Commit:
i1030: remove debugging prints.

Commit:
i1030: add GroupListener; update AccountListener.

Commit:
i1030: update Group and Team change listeners; also...

Only show group/team panes if they have data.
Update documentation, remove old unused code.

Commit:
i1030: add getContestStartTime and toString to ContestTimeImpl; note:

This was ported from the i1027_Add_Contest_Clock_To_WTI branch.

Commit:
i1030: add getContestStartTime() to interface IContestClock.  Note:

This was ported from the i1027_Add_Contest_Clock_To_WTI branch.

Commit:
i1030: import WTI-API ContestClockModel from branch i1027.

Commit:
i1030: add /contestClock endpoint.  Note that this also includes...

...adding ContestClockModel and updated IContestClock.
These were added in order to support allowing the /clarifications
endpoint to return clars any time (previously, clars were only returned
if the contest was RUNNING), and allowing the /problems endpoint to
return the problem list any time after the contest has been started
(previously, problems were only being returned if the contest was
RUNNING).

These updates were ported into this branch from the
i1027_Add_Contest_Clock_To_WTI branch.

Commit:
i1030: minor variable renaming for clarity; no logic changes.

Commit:
i1030: add destination teams and groups lists to ClarImpl.

Commit:
i1030: add clar destination teams & groups to WTI ClarificationModel.

Commit:
i1030: remove (comment-out) WTI-UI clar filter-by-recipient buttons.

Commit:
i1030: remove (comment-out) filtering of clars.

Commit:
i1030: insert team and group destinations into clars sent to WTI-UI.

Commit:
i1030: restore filter buttons on WTI-UI Clarifications page.

Commit:
i1030: restore clar filtering; include "not answered" in all displays.

Commit:
i1030: support displaying "not answered" clars in WTI-UI.

Commit:
i1030: clear selected groups/teams when switching to "All Teams" so...

... that the user doesn't accidentally send to "selected groups and
teams" after switching back to "all teams" because any
previously-selected groups/teams ARE STILL SELECTED even though the
scrollpanes are invisible.

clevengr committed:
i1030: fix bug in getAllDestinationGroups/Teams().  Specifically...

these methods were calling getFirstAnswer() to get the destination
groups/teams -- but that throws an NPE in the case of a clarification
which hasn't yet been answered.

clevengr committed:
i1030: update copyright date in shared footer.

kkarakas commented:
My understanding is that this PR builds on top of WTI Clock PR. Therefore it's not apt to make code review or functionality review without merging that one first. After WTI clock is merged I can look at this.

clevengr commented:
> My understanding is that this PR builds on top of WTI Clock PR. Therefore it's not apt to make code review or functionality review without merging that one first. After WTI clock is merged I can look at this.

No, this PR is not dependent on the WTI Clock PR (#1034). The branch on which this PR was based was created directly off the `develop` branch and depends only on that branch.  (I did merge a few pieces of code from the WTI clock branch into the code of this PR -- but that code is part of this PR and doesn't make this PR dependent on the WTI clock PR...)

I think this PR is ready for review, independent of merging any other PR.  If you find something that explicitly contradicts this, please let me know.

kkarakas reviewed (commented):
Did not review the code. I did review the functionality; it looks great works as expected. There are 3 points that I would like to discuss:

1. Announcement Textbox jumps up and down if you switch back and forth the Announcement Destination which I don't think looks elegant. 
2. In Announcement pane, if you select the problem, if you select the destination and if you type a announcement text but do not select any groups or teams and press the button "Send Announcement" judge/admin throws an exception. I think judge/admin should warn the user that they forgot to select a group and/or team. 
3. I wonder if judge/admin should clear all the selected groups and/or teams after announcement is broadcasted. I feel like it is unlikely they will send the announcement to same exact destinations.

kkarakas commented on projects/WTI-API/src/main/controllers/ContestController.java:
```diff
@@ -439,22 +552,15 @@ public Response clarifications(@ApiParam(value="token used by logged in users to
 		ServerConnection userInformation = connections.get(key);
 	    
-		//verify the user is logged in and the contest is running
+		//verify the user is logged in
```
"verify that the user is logged in" Would be a better sentence to describe the situation.

clevengr replied:
Fixed in next push.

kkarakas commented on src/edu/csus/ecs/pc2/api/implementation/ContestTimeImplementation.java:
```diff
@@ -1,6 +1,8 @@
 // Copyright (C) 1989-2019 PC2 Development Team: John Clevenger, Douglas Lane, Samir Ashoo, and Troy Boudreau.
```
copyright could be changed.

clevengr replied:
Fixed in next push.

kkarakas commented on src/edu/csus/ecs/pc2/core/InternalController.java:
```diff
@@ -3892,7 +3892,7 @@ public ElementId submitClarification(Problem problem, String question) {
     public void submitAnnouncement(Problem problem, String answer,ElementId[] ultimateDestinationGroup, ClientId[] ultimateDestinationTeam) {
         ClientId serverClientId = new ClientId(contest.getSiteNumber(), Type.SERVER, 0);
-        Clarification clarification = new Clarification(contest.getClientId(), problem, "");
+        Clarification clarification = new Clarification(contest.getClientId(), problem, "Announcement");
```
Wouldn't this be problematic if a team submits a clarification with "Announcement" text? The clarification would be interpreted as announcement and break the system.

clevengr replied:
No, it's not a problem.
I just did the following tests to verify that it works correctly: 
- I submitted a Clarification (from a WTI team) whose text was exactly "Announcement".
- I answered the clar, sending the answer to just the initiating team.

The result was that the initiating team saw a clarification whose "question" was "Announcement", and whose "Answer" was the text I sent back (from the Admin) as an answer.
- I then repeated the above, but sending the answer to all teams.

The result was as expected:  all teams (including the originating team) get a clar whose "question" is "Announcement" and whose answer is whatever the Admin/Judge answers. 

The reason it's not a problem is that the code which you commented on (above)  is in method `InternalController.submitAnnouncement()`, which is only called from one single place:  `ui.GenerateAnnouncementPane.submitAnnouncement()`.  It is the InternalController `submitAnnouncement()` method  that constructs a "Clarification" whose text is set to "Announcement".  But there's no way for a user (i.e. the WTI) to invoke that method; when a user submits a clarification (even if the "question" in the clarification is "Announcement"), it doesn't invoke `InternalController.submitAnnouncement()` -- it invokes `InternalController.submitClarification()`.  The result is that a Clarification whose text is "Announcement" is handled just like any other clarification; it doesn't cause any conflict with an actual Announcement just because the Announcement has had it's "question" text set to "Announcement".

It is true that IF a team submits a clarification whose question text is "Announcement" then they will see the word "Announcement" in the "Question" column on their display, just as they would see the word "Announcement" in that column for an actual announcement.   However, I don't see this as a problem -- first because the team will know that they submitted such a (bizarre) question in the first place, and second because it seems like the Judges would provide some kind of appropriate _answer_ to such a (non-sensical) "question".   IMHO this is a small price to pay for what is gained -- all _actual_ announcements from the judges will contain the word "Announcement" in the question field.  I think that helps teams more than just having the question field be blank.

kkarakas reviewed (commented):
These are my reviews.

Commit:
i1030: use Box.glue to produce better vertical spacing

clevengr committed:
i1030: Updates for PR changes requested by Kutay. Specifically:

-Fixed typos and updated copyright date in some comments.
-Replaced some struts with springs (aka glue) in the
GenerateAnnouncementPand layout, to help reduce the jumping of the
Announcement Text pane.
-Prompts user when "Send Announcement" is pushed but no Groups or Teams
have been selected.
-Also: numerous minor CI fixes.

clevengr commented:
> Did not review the code. I did review the functionality; it looks great works as expected. There are 3 points that I would like to discuss:
> 
> 1. Announcement Textbox jumps up and down if you switch back and forth the Announcement Destination which I don't think looks elegant.

I agree this is slightly inelegant.  However, it's a complex LayoutManager issue.  I spent hours trying to fix it, including approaches such as:
- Leaving the Groups/Teams pane visible (instead of hiding them) but disabling the checkboxes.  I couldn't get that to work.
  - (Specifically, what I tried was disabling the checkboxes in the model. However, this did not clear the checked boxes on the screen.  It MAY be that all that was still needed is a "repaint()" of the panel -- but I ran out of time to try that, and anyway I'm not necessarily conviced it would solve the problem.)
- Adding various combinations of struts and springs (a.k.a. "glue") between various components.

None of these approaches completely solved the problem.  I think the only real solution is to completely re-layout the GUI, perhaps using `GridBagLayout`.  I don't feel like I have the time to do that right now, so this should be deferred to a new Issue.
What I DID do was replace some of the fixed "struts" in the layout with springs (glue).  That seems to help reduce the jumping a little bit...

> 2. In Announcement pane, if you select the problem, if you select the destination and if you type a announcement text but do not select any groups or teams and press the button "Send Announcement" judge/admin throws an exception. I think judge/admin should warn the user that they forgot to select a group and/or team.

I completely agree; nice catch.  Fixed in the next push - it now prompts the user to select a group or team.

> 3. I wonder if judge/admin should clear all the selected groups and/or teams after announcement is broadcasted. I feel like it is unlikely they will send the announcement to same exact destinations.

I could go either way on this, so I accepted your suggestion and implemented this functionality in the next push.

Commit:
i1030: fix conflicts with merge into develop

Commit:
i1030: fix conflicts with PR merge

clevengr commented:
This branch still has merge conflicts with the develop branch.  I have looked at all the conflicts in the two conflicting files; they are explainable and I know how to fix them.  However, I will wait to do that until the PR itself is approved.  Waiting for reviews/approvals before fixing and merging.

kkarakas reviewed (approved):
Everything looks great, works as expected and previous issues that I talked about were addressed.

johnbrvc commented on src/edu/csus/ecs/pc2/ui/JPanePlugin.java:
```diff
@@ -381,26 +381,44 @@ public static void showExceptionMessage(Component component, final String messag
-            if (i < groupIds.length) {
+        //check if we have any groups (the above test only returns if BOTH groups AND teams are null...)
+        if (groupIds != null) {
+            //we have some groups; append each group name to the string, with comma separators
```
Actually, it uses "comma-space" separators.

johnbrvc reviewed (commented):
Code review complete.  I made one very minor comment.
I looked over the code under "projects", but I'm not up on the architecture of WTI so in terms of functionality, I couldn't comment.  Code seems OK.  

The test procedure is very long and time consuming - between code review and setting up the test environment I have spend over 2 hours so far - and - to be honest, I have very little idea how the code works or what is going on under the hood.

I have begun the testing procedure but it will take some time to go through ALL those test cases.

If the author is satisficed that all those cases work, I have no problem merging this before I get to all the test cases.  We have to weigh the possible consequences if something is wrong, would it be a show stopper.

clevengr commented:
> ...
> I have begun the testing procedure but it will take some time to go through ALL those test cases.
> 
> If the author is satisficed that all those cases work, I have no problem merging this before I get to all the test cases. We have to weigh the possible consequences if something is wrong, would it be a show stopper.

I (the author) am pretty confident that all the specified test cases work.  I just re-executed them all (to make sure they still work following the most recent code updates.  I think it's worth moving to merging so that we can see what conflicts will arise.

clevengr committed:
Merge branch 'develop' into i1030_WTI_Clar_Announcement_Filtering
